rclcpp/include/rclcpp/generic/generic_publisher.hpp

```diff
@@ -0,0 +1,92 @@
+// Copyright 2018, Bosch Software Innovations GmbH.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef RCLCPP__GENERIC__GENERIC_PUBLISHER_HPP_
+#define RCLCPP__GENERIC__GENERIC_PUBLISHER_HPP_
+
+#include <memory>
+#include <string>
+
+#include "rclcpp/publisher_base.hpp"
+#include "rclcpp/node_interfaces/node_base_interface.hpp"
+#include "rclcpp/node_interfaces/node_topics_interface.hpp"
+#include "rclcpp/macros.hpp"
+#include "rclcpp/qos.hpp"
+#include "rclcpp/callback_group.hpp"
+#include "rcpputils/shared_library.hpp"
+
+namespace rclcpp_generic
+{
+
+/**
+ * This class is an implementation of an rclcpp::PublisherBase for serialized messages whose topic
+ * is not known at compile time (hence templating does not work).
+ *
+ * It does not support intra-process handling.
+ */
+class GenericPublisher : public rclcpp::PublisherBase
+{
+public:
+  // cppcheck-suppress unknownMacro
+  RCLCPP_SMART_PTR_DEFINITIONS(GenericPublisher)
+
+  /**
+   * Factory method.
+   *
+   * The returned pointer will never be empty, but this function can throw various exceptions, for
+   * instance when the message's package can not be found on the AMENT_PREFIX_PATH.
+   *
+   * \param topics_interface NodeTopicsInterface pointer used in parts of the setup
+   * \param topic_name Topic name
+   * \param topic_type Topic type
+   * \param qos QoS settings
+   * \param group Callback group
+   */
+  static std::shared_ptr<GenericPublisher> create(
+    rclcpp::node_interfaces::NodeTopicsInterface::SharedPtr topics_interface,
+    const std::string & topic_name,
+    const std::string & topic_type,
+    const rclcpp::QoS & qos,
+    rclcpp::CallbackGroup::SharedPtr group = nullptr);
```

Comment:
Thanks, yeah, that's what I was wondering about this factory function too. Is there any reason for this free function pattern?

Reply:
I noticed the other free `create` function take a `NodeBaseInterface` in addition to `NodeTopicsInterface`/`NodeServicesInterface`, but I didn't because it's possible to obtain from `NodeTopicsInterface::get_node_base_interface()` – is there a reason for having both interfaces or is it fine to leave only the topics interface in?

Reply:
> Is there any reason for this free function pattern?

Yes, the reason is so that other types of nodes can use these functions, e.g. LifecycleNode:

```
auto node = std::make_shared<MyCustomNode>(...);
auto pub = rclcpp::create_publisher<...>(node, ...);
```

Without having a method themselves. Or they can use the free function to implement their version of the method without duplicating boilerplate.

Normally this would be the constructor of the things we're making, but for other reasons those constructors need to be protected.

> is there a reason for having both interfaces or is it fine to leave only the topics interface in?

No this is basically legacy. You can pass just the `NodeTopicsInterface` and obtain the `NodeBaseInterface` from it.

Reply:
I made the change. My constructors don't need to be protected, do they? There is also `NodeTopics::create_subscription` (but no `NodeServices::create_client`) – should I add that as well?

Reply:
> My constructors don't need to be protected, do they?

I'd say no?

> There is also NodeTopics::create_subscription (but no NodeServices::create_client) – should I add that as well?

Yes, maybe. Those functions do two things, they add the entity (sub or client or whatever) to the callback group so it is kept track of in the node, and then also they notify the system that something was added to the callback group (and therefore to the node). This is used, for example, by the executor so it knows it needs to stop waiting and reevaluate what to wait on because there's new things that it may need to consider.